docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+by the core provider.
+
+##### Upgrades providers without changing contract
+
+If the new version of the provider does abide by the same version of the
+Cluster API contract, the operator will execute the upgrade by performing:
+- Delete of the current instance of the provider components, while preserving
+  CRDs, namespace and user objects.
+- Install the new version of the provider components
+
+Please note that:
+- The operator executes upgrades 1 provider at time, while `clusterctl upgrade
+  apply` manages upgrading a group of providers with a single operation.
+- `clusterctl upgrade apply --contract` automatically determines the latest
+  versions available for each provider, while with the Declarative approach
+  the user is responsible for manually editing Provider objects yaml.
+- `clusterctl upgrade apply` currently uses environment variables and a local
+  configuration file; this should be changed in order to use in cluster
+  provider configurations.
+
+![Figure 1](./images/management-cluster-operator/fig1.png "Figure for
+upgrading provider without changing contract")
+<div align="center">Upgrading providers without changing contract</div>
+<br/>
+
+##### Upgrades providers and changing contract
+
+If the new version of the provider does abide by a new version of the Cluster
+API contract, it is required to ensure all the other providers in the
+management cluster should get the new version too.
+
+![Figure 2](./images/management-cluster-operator/fig2.png "Figure for
+upgrading provider and changing contract")
+<div align="center">Upgrading providers and changing contract</div>
+<br/>
+
+As a first step, it is required to pause all the providers by setting the
+`spec.Paused` field to true for each provider; the operator will block any
+contract upgrade until all the providers are paused.
+
+After all the providers are in paused state, you can proceed with the upgrade
+as described in the previous paragraph (change the `spec.Version` field).
+
+Once all the providers are upgraded to a version that abides to the new
+contract, it is possible for the operator to unpause providers; the operator
+does not allow to unpause providers if there are still providers abiding to
+the old contract.
+
+Please note that we are planning to embed this sequence (pause - upgrade -
+unpause) as a part of `clusterctl upgrade apply` command when there is a
+contract change.
+
+
+##### Changing a provider
+
+On top of changing a provider version (upgrades), the operator supports also
+changing other provider fields, most notably controller flags and variables.
+
+The operation internally works like upgrades: The current instance of the
+provider is deleted, while preserving CRDs, namespaced and user objects A new
+instance of the provider is installed with the new set of flags/variables.
+
+Please note that clusterctl currently does not support this operation.
+
+See Example 1 in [Example API Usage](#example-api-usage)
```

Comment:
what about expanding this a little by explaining that the user is required to kubectl edit (or to kubeactl apply again) the object describing the provider